File: 1/2020.09.12/142.Linked List Cycle II/python/solution.py
class Solution:
    def detectCycle(self, head: ListNode) -> ListNode:
        slow = head
        fast = head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            if slow is fast:
                begin = head
                while begin is not slow:
                    begin = begin.next
                    slow = slow.next
                return begin
        return None


    def detectCycle2(self, head: ListNode) -> ListNode:
        try:
            slow = head
            fast = head.next
            while slow is not fast:
                slow = slow.next
                fast = fast.next.next
            slow = slow.next
            begin = head
            while begin is not slow:
                begin = begin.next
                slow = slow.next
            return begin
        except:
            return None

File: 1/2020.09.12/142.Linked List Cycle II/java/solution.java
public class Solution {
    public ListNode detectCycle(ListNode head) {
    	ListNode slow = head;
		ListNode fast = head;
		while (fast != null && fast.next != null) {
			slow = slow.next;
			fast = fast.next.next;
			if (slow == fast) {
				ListNode begin = head;
				while (begin != slow) {
					begin = begin.next;
					slow = slow.next;
				}
				return begin;
			}
		}
		return null;
    }
    
    public ListNode detectCycle(ListNode head) {
    	try {
            ListNode slow = head;
            ListNode fast = head.next;
            while (slow != fast) {
                slow = slow.next;
                fast = fast.next.next;
            }
            slow = slow.next;
            ListNode begin = head;
            while (begin != slow) {
                begin = begin.next;
                slow = slow.next;
            }
            return begin;
        } catch (NullPointerException e) {
            return null;
        }
    }
